File: 6 – Practice/leet code/003. Longest Substring Without Repeating Characters.cpp
// https://leetcode.com/problems/longest-substring-without-repeating-characters/
// Sliding window

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        const int SIZE = s.size();
        int lo = 0;
        int hi = 0;
        
        int mxlen = 0;
        
        const int ASCII = 256;
        int used[ASCII] = {0};
        
        while(hi < SIZE) {
            if(used[s[hi]]) {
                lo = max(lo, used[s[hi]]);
            }
            
            used[s[hi]] = hi + 1;
            mxlen = max(mxlen, hi - lo + 1);
            
            ++hi;
        }
        
        return mxlen;
    }
};

File: 6 – Practice/leet code/003. Longest Substring Without Repeating Characters.java
// https://leetcode.com/problems/longest-substring-without-repeating-characters/
// Smart sliding window

class Solution {
    public int lengthOfLongestSubstring(String s) {
        final int SIZE = s.length();

        int lo = 0;
        int hi = 0;
        int mxlen = 0;
       
        final int ASCII = 256;

        int[] used = new int[ASCII];
        while (hi < SIZE) {
            if (used[s.charAt(hi)] != 0) {
                lo = Math.max(lo, used[s.charAt(hi)]);
            }

            used[s.charAt(hi)] = hi + 1;
            mxlen = Math.max(mxlen, hi - lo + 1);

            hi++;
        }

        return mxlen;
    }
}
